docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+  - In a consistent state images should move in the same direction or stay in
+    place. This means in practice that `swap_type` cannot be at the same time
+    `REVERT` and `<PERM | TEST>`.
+  - These `swap_type` combinations are valid:
+    - `NONE` or `<PERM | TEST>`
+    - `REVERT` or `NONE`
+
+# Testing in simulator
+Simulator environment must be improved to able to test possible scenarios during
+multiple image update, focusing on abort and revert scenarios and on the usage
+of common scratch area.
+
+# Testing on hardware
+Testing can be done on ARMv8-M based platform i.e.: MPS2_AN521 or Musca A1.
+MCUBoot can be compiled without knowing that it is running on ARMv8-M. Code,
+which was compiled to ARMv7-M is compatible with ARMv8-M. Just the memory map
+must be correct. Also the compiler does not need to support ARMv8-M
+architecture.
+
+Test application, specially the secure code must be compiled with ARMv8-M
+awareness. Currently this can be done in TF-M project.
+
+Testing on ARMv7-M based platform is also possible. Just need to compile two
+application with different and not overlapping memory map. They can be executed
+together. The first application can be started by the bootloader and second
+application is started by the first one, similar to secure and non-secure code
+on ARMv8-M.
+
+# Alternative solution
+Use only one secondary area and upgrade the images one-by-one after each other.
+Size of secondary area must be as big as the biggest image primary area.
+Benefits:
+  - Reduced the complexity of the code, because this is basically the currently
+    used process with single image update.
+  - Reduced flash fragmentation
+  - Better utilization of flash
+Drawbacks:
+  - When API is changing between images then the first image which is updated
+    it must be able to handle both version of the API to can interact with still
+    old version of other image. Then other image is updated it, it needs to
+    handle only the new version of API.
+  - It makes updates more complex, in some cases. But if API change is a rear
+    event and usually independent updates can be done then this approach is
+    feasible.
+
+# Unresolved questions
+  - How the multiple image update conflicts with split image update?
+  - How to combine with rollback protection? After a revert restrict old image
+    functionality to a minimal subset to be able to receive new fixed version or
+    run some diagnostic, but main functionality is disabled.
+  - It was assumed that the result of self test, which sets the `image_ok` flag,
+    is independent from the other image. In theory it means if both image was
+    updated, but one of them must be reverted then after the revert the
+    previously successful self test is expected to be still successful
+    regardless that other image was reverted. This might be not the case in
+    practice.
+  - Does it depend anyway on the RTOS (Mynewt or Zephyr)?
+  - Serial recovery with multiple images.
```

Comment:
This will require updating `mcumgr` and `newtmgr` and could probably be done later in another task. You could check that `BOOT_SERIAL` cannot be defined together with `DUAL_IMAGE_SUPPORT` for now.

Reply:
What does it mean, then to revert one of the images? What if the upgrade was atomic, we then need to make sure that both are reverted.

Honestly, I think we should revisit the auto-rollback requirement, and perhaps determine if that is needed, or even should be implemented in the multiple-image case.

Reply:
Whether the upgrade is atomic or not it depends on dependency entries. Atomic update, when new images must upgraded and reverted together, it can be enforced with strict dependencies. But if the dependencies are loose, there were no API change, so it remains the old value. Then still both image can be updated, and if the BIST is failing on one of them in that case only the faulty image will be reverted. 

My original assumption was that the image's BIST result are independent from the other image, but this might be not the best approach. BIST should test the working system and the flags must be in sync:
- If both image was updated then after BIST both `image_ok` flag should be OK, or NOK.
- If only one of the image was updated then only the corresponding `image_ok` must be set, and the other should not changed.